Add timezone field to Courses object #4812

wkurniawan07 opened the pull request:
Fixes #4812 

This continues the work done in #4813. So far only the works from that PR plus de-conflicting, fixing static analysis errors, and updating the libraries (moment, selectize) are done. Will still be a while before ready for review.

wkurniawan07 commented:
We have a functional `timeZone` field now, tested by CRUD-ing new courses as well as RUD-ing old courses. Since the field is not yet used anywhere, its presence/absence is still of decorative purpose as of now.
The next courses of action will be migrating the timezone field from session to course. Whether or not the said field can be retained in feedback sessions is another point of discussion, but migration will still be needed because the sessions do not store time zone; rather they store time zone **offsets** (this is precisely why any attempt to handle DST in our current condition will result in epic failure).

wkurniawan07 commented:
@JosephineKwa I recall you reviewed the previous PR touching this matter 😛

wkurniawan07 commented on src/test/java/teammates/test/cases/ui/InstructorCourseEditSaveActionTest.java:
```diff
@@ -132,6 +138,26 @@ public void testExecuteAndPostProcess() throws Exception {
+                            FieldValidator.REASON_UNAVAILABLE_AS_CHOICE);
+        assertEquals(statusMessage, redirectResult.getStatusMessage());
+        assertEquals(Const.ActionURIs.INSTRUCTOR_COURSE_EDIT_PAGE
+                     + "?error=true&user=" + instructorId + "&courseid=" + courseId,
```
Oops wrong indentation 😅

wkurniawan07 commented on src/test/java/teammates/test/driver/HtmlHelper.java:
```diff
@@ -159,6 +159,14 @@ private static String convertElementNode(Node currentNode, String indentation, b
                     return generateStudentMotdPlaceholder(indentation);
                 }
             }
+        } else if (currentNode.getNodeName().equalsIgnoreCase("select")) {
```
This is to prevent all the timezone options from being printed in the HTML test files, otherwise the LoCs here will reach ~1800 * number of test files affected (can easily reach 20000). This may or may not be temporary.

wkurniawan07 commented on src/test/resources/pages/allJsUnitTests.html:
```diff
@@ -31,7 +31,6 @@
   <!-- Test scripts -->
   <script type="text/javascript" src="../../javascript/CommonJsTest.js"></script>
   <script type="text/javascript" src="../../javascript/InstructorCourseDetailsJsTest.js"></script>
-  <script type="text/javascript" src="../../javascript/InstructorCoursesJsTest.js"></script>
```
The said test file no longer exists.

wkurniawan07 commented:
@JosephineKwa ping here.

JosephineKwa commented on src/main/webapp/WEB-INF/tags/instructor/course/courseEditCourseInfo.tag:
```diff
@@ -49,6 +49,21 @@
+                    </select>
+                </div>
+                <div class="col-sm-1">
+                    <input class="btn btn-primary" id="auto-detect-timezone" disabled value="Auto-Detect Timezone">
```
Add a type="button" here so that the input button won't be editable :p

JosephineKwa replied:
Also to standardize things, perhaps add a space in `Timezone` in the value, as well as a hyphen in `timezone` in the id

wkurniawan07 replied:
There are close-to-equal spread of `timezone`/`TIMEZONE` and `timeZone`/`time-zone`/`TIME_ZONE` throughout the code base so I don't think this is a big priority for now.

JosephineKwa commented on src/main/java/teammates/common/datatransfer/CourseAttributes.java:
```diff
@@ -34,26 +34,31 @@ public int compare(CourseAttributes course1, CourseAttributes course2) {
     
-    public CourseAttributes(String courseId, String name, boolean archiveStatus) {
+    public CourseAttributes(String courseId, String name, String timeZone,
+                            boolean archiveStatus) {
```
Could shift this to the line before

JosephineKwa commented on src/main/webapp/js/instructorCourseEdit.js:
```diff
@@ -462,4 +469,14 @@ $(document).ready(function() {
+        TimeZone.updateTimeZone($selectElement, courseTimeZone);
+
+        $('#auto-detect-timezone').on('click', function() {
+            autoDetectTimezone();
```
If we are standardize time zone to have a space, this can be renamed as `autoDetectTimeZone();`

JosephineKwa commented on src/main/java/teammates/common/util/Const.java:
```diff
@@ -63,6 +65,7 @@
     
     public static final String DEFAULT_SECTION = "None";
     
+    public static final String DEFAULT_TIMEZONE = DateTimeZone.UTC.getID();
```
Also add an underscore in `TIMEZONE`

JosephineKwa commented on src/main/webapp/jsp/instructorCourseEdit.jsp:
```diff
@@ -5,7 +5,13 @@
 <c:set var="jsIncludes">
+    <script type="text/javascript" src="/js/lib/moment.min.js"></script>
+    <script type="text/javascript" src="/js/lib/moment-timezone-with-data-2010-2020.min.js"></script>
+    <script type="text/javascript" src="/js/timezone.js"></script>
```
Could also consider renaming the file to `timeZone.js`

JosephineKwa commented on src/test/java/teammates/test/cases/storage/CoursesDbTest.java:
```diff
@@ -71,6 +71,17 @@ public void testCreateCourse() throws EntityAlreadyExistsException, InvalidParam
                                         e.getMessage());
         }
 
+        CourseAttributes invalidTimezoneCourse =
```
Here too :stuck_out_tongue_closed_eyes:

JosephineKwa commented:
Changes look good overall, some changes to address!

wkurniawan07 commented:
Addressed the comments, also added the script to populate the course timezones from its existing sessions (I refrain from using "migrate" because the timezone field in sessions is not removed and still used, and the timezone field in courses is still, like I put it before, decorative).

wkurniawan07 commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,135 @@
+        map("12.0", "Pacific/Auckland");
+        map("12.75", "Pacific/Chatham");
+        map("13.0", "Pacific/Tongatapu");
+        map("14.0", "Pacific/Kiritimati");
```
All these values are arbitrarily chosen, although there are obvious biases such as `Asia/Jakarta` and `Asia/Singapore` :p

JosephineKwa commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,136 @@
+        try {
+            course.setTimeZone(timeZone);
+            logic.updateCourse(course);
+            System.out.println("Course + " + course.getId() + " timezone successfully set to " + timeZone + ".");
```
Is the `+` in `"Course + "` intended? Perhaps it could be removed

JosephineKwa commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,136 @@
+            logic.updateCourse(course);
+            System.out.println("Course + " + course.getId() + " timezone successfully set to " + timeZone + ".");
+        } catch (InvalidParametersException | EntityDoesNotExistException e) {
+            System.out.println("Failed to set timezone for course + " + course.getId() + ".");
```
The `+` after `course` here could be removed too

JosephineKwa commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,136 @@
+    }
+    
+    private List<CourseAttributes> getAllCoursesWithoutTimeZone() {
+        List<CourseAttributes> courseAttributes = new ArrayList<CourseAttributes>();
```
Maybe name this as `courses` as `courseAttributes` could refer to a single instance of `CourseAttributes`

JosephineKwa commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,136 @@
+    private List<CourseAttributes> getAllCoursesWithoutTimeZone() {
+        List<CourseAttributes> courseAttributes = new ArrayList<CourseAttributes>();
+        List<Course> courseEntities = getAllCourseEntities();
+        for (Course course : courseEntities) {
```
If the above suggestion is taken, `course` could be renamed as `courseEntity` to avoid a name clash with `courses` which represents a list of `CourseAttributes`

JosephineKwa commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,136 @@
+                                             : getTimeZoneId(sessions.get(0).getTimeZone());
+        
+        if (isPreview) {
+            System.out.println("Course " + course.getId() + " to be set to " + timeZone + ".");
```
Maybe add `"timezone"` after `course.getId()`

JosephineKwa commented:
Looks good!

damithc commented:
I didn't check the code yet, but I would like to ask if these aspects have been considered.
1.  Deploying the code and data migration will not happen simultaneously. The new code should continue to work even if the old courses have not been migrated. 
2. How do you ensure sample courses created match the user time zone?

wkurniawan07 commented:
1. The code works regardless of presence of timezone in courses in database.
2. For existing courses: I'm afraid there's no easy way for that. For new courses (after this change is deployed): This change alone doesn't make it possible. It can be done during instructor account creation (meaning the account request form needs this change as well). However, it should be noted that even the current existing sample sessions don't take into account the user's timezone.

damithc commented:
I'll review the code soon.

damithc commented on src/client/java/teammates/client/scripts/DataMigrationForTimezoneFieldInCourses.java:
```diff
@@ -0,0 +1,136 @@
+    private void updateTimezoneForCourse(CourseAttributes course) {
+        List<FeedbackSessionAttributes> sessions = logic.getFeedbackSessionsForCourse(course.getId());
+        String timeZone = sessions.isEmpty() ? Const.DEFAULT_TIMEZONE
+                                             : getTimeZoneId(sessions.get(0).getTimeZone());
```
1. Should we leave the timeZone as 'unset' so that it can be set to the correct value later at the first possible opportunity? If the code can do that, the same strategy can be used for sample courses.
2. I'm not sure if String is the right type here. The Date library we are using doesn't have a more suitable type? If not, should we define an enum/class for it?

wkurniawan07 replied:
1. Possible currently, but only until we start to use the timezone data from the courses object. After which it is non-nullable anymore. So we might as well disallow null value from the start, i.e during migration/population.
2. Nope, only String, because it's only storing ID and nothing else. Enum is not feasible because there are ~600 timezone IDs and they are not static in that updates to libraries can add new IDs or remove some old ones.

damithc commented:
Waiting for some clarifications. Meantime, can post screenshots of any UI that changed?

wkurniawan07 commented:
Only course add and edit pages are affected.
![screen shot 2016-08-05 at 11 44 04 pm](https://cloud.githubusercontent.com/assets/7261051/17442449/65a51d9a-5b68-11e6-951d-eef3dc8e25a9.png)
![screen shot 2016-08-05 at 11 44 15 pm](https://cloud.githubusercontent.com/assets/7261051/17442554/d715ae90-5b68-11e6-8b1d-fefcfdf47709.png)
![screen shot 2016-08-05 at 11 44 50 pm](https://cloud.githubusercontent.com/assets/7261051/17442567/defffd4a-5b68-11e6-89ce-124f22d5c5a5.png)

damithc commented:
Data migration completed.
